app/handlers/tests/test_handlers_services.py

```diff
@@ -0,0 +1,184 @@
+from unittest.mock import MagicMock, patch
+
+import kopf
+import kubernetes
+import pytest
+import yaml  # type: ignore
+
+from app.handlers.handlers_services import (
+    k8s_get_twingate_resource,
+    service_to_twingate_resource,
+    twingate_service_create,
+)
+
+# Ignore the fact we use _cogs here
+
+
+@pytest.fixture()
+def example_service_body():
+    yaml_str = """
+    apiVersion: v1
+    kind: Service
+    metadata:
+      name: my-service
+      annotations:
+        twingate.com/expose: "true"
+        twingate.com/expose-alias: "myapp.internal"
+    spec:
+      selector:
+        app.kubernetes.io/name: MyApp
+      ports:
+        - name: http
+          protocol: TCP
+          port: 80
+          targetPort: 9376
+        - name: https
+          protocol: TCP
+          port: 443
+          targetPort: 9377
+        - protocol: UDP
+          port: 22
+          targetPort: 9376
+          name: ssh
+    """
+    return kopf._cogs.structs.bodies.Body(  # noqa: SLF001
+        yaml.safe_load(yaml_str)
+    )
+
+
+@pytest.fixture()
+def k8s_customobjects_client_mock():
+    client_mock = MagicMock()
+    with patch("kubernetes.client.CustomObjectsApi") as k8sclient_mock:
+        k8sclient_mock.return_value = client_mock
+        yield client_mock
+
+
+def test_service_to_twingate_resource_with_alias(example_service_body):
+    result = service_to_twingate_resource(example_service_body, "default")
+
+    assert result == {
+        "apiVersion": "twingate.com/v1beta",
+        "kind": "TwingateResource",
+        "metadata": {
+            "name": "my-service-resource",
+        },
+        "spec": {
+            "name": "my-service-resource",
+            "address": "my-service.default.svc.cluster.local",
+            "alias": "myapp.internal",
+            "protocols": {
+                "allowIcmp": False,
+                "tcp": {
+                    "policy": "RESTRICTED",
+                    "ports": [{"start": 80, "end": 80}, {"start": 443, "end": 443}],
+                },
+                "udp": {
+                    "policy": "RESTRICTED",
+                    "ports": [{"start": 22, "end": 22}],
+                },
+            },
+        },
+    }
+
+
+def test_service_to_twingate_resource_without_alias(example_service_body):
```

Comment:
nit: Would be nice to create test case for `service_to_twingate_resource` function to group this and the test above.